working/0015-infer-required/feature-specification.md

```diff
@@ -0,0 +1,213 @@
+*    [#2989 Purpose of required before named non-nullable arguments](https://github.com/dart-lang/language/issues/2989)
+*    [#3206 Rethink required to be optional for non-nullable named parameters without default values](https://github.com/dart-lang/language/issues/3206)
+*    [#3287 Inferring required named parameters without making function types a pitfall](https://github.com/dart-lang/language/issues/3287)
+
+## Motivation
+
+For brevity, it is desirable to be able to omit the rather long modifier
+`required` on the declaration of a named formal parameter. Of course, the
+resulting declaration should still be unambiguous, but this is indeed
+possible in some cases.
+
+This section mentions a different proposal as well, namely
+[primary constructors](https://github.com/dart-lang/language/blob/main/working/2364%20-%20primary%20constructors/feature-specification.md).
+The reason for this is that this proposal about eliminating `required`
+turns out to be even more significant when the underlying declarations are
+concise, and primary constructors will do just that.
+
+Note that all parameters in this document are named, because the proposal
+is specifically concerned with the rules about named parameters.
+
+For example:
+
+```dart
+// Current form.
+
+class Point {
+  final int x;
+  final int y;
+  const Point({required this.x, required this.y});
+}
+
+// If this proposal is supported.
+
+class Point {
+  final int x;
+  final int y;
+  const Point({this.x, this.y});
+}
+
+// If primary constructors are supported.
+
+class const Point({required int x, required int y}); 
+
+// With primary constructors plus this proposal.
+
+class const Point({int x, int y});
+```
+
+Here is a larger example (from issue 878):
+
+```dart
+// Today.
+
+class User {
+  final String id;
+  final String email;
+  final String userName;
+  final String address;
+  final String phoneNumber;
+  final String name;
+  final String? avatarUrl;
+
+  User({
+    required this.id,
+    required this.email,
+    required this.userName,
+    required this.address,
+    required this.phoneNumber,
+    required this.name,
+    this.avatarUrl,
+  });
+}
+
+// With this proposal.
+
+class User {
+  final String id;
+  final String email;
+  final String userName;
+  final String address;
+  final String phoneNumber;
+  final String name;
+  final String? avatarUrl;
+
+  User({
+    this.id,
+    this.email,
+    this.userName,
+    this.address,
+    this.phoneNumber,
+    this.name,
+    this.avatarUrl,
+  });
+}
+
+// With this proposal and primary constructors.
+
+class const User({
+  String id,
+  String email,
+  String userName,
+  String address,
+  String phoneNumber,
+  String name,
+  String? avatarUrl,
+});
+```
+
+It seems likely that it will be true quite often that the required, named
+parameters are exactly the ones whose type is potentially non-nullable,
+which means that the migration will simply be to delete every occurrence of
```

Comment:
If we have an automated migration I think we should make give it full fidelity - a `required` which would not be inferred shouldn't be removed.
I don't know of specific examples, but I think it's reasonable for existing code to use `required` to force callers to explicitly pass `null` when that is the intention.

Reply:
Right, a mechanical migration should certainly be semantics preserving. Added a sentence to clarify this point.